File: src/modules/klattgrid_module.cpp
// klattgrid_module.cpp
// Rcpp Module for Praat KlattGrid object (speech synthesis)
// Phase 2.3 - Articulatory speech synthesis with Klatt formant synthesizer

#include <Rcpp.h>
#include "module_common.h"

// Praat headers
#include "../praat.github.io/dwtools/KlattGrid.h"
#include "../praat.github.io/fon/Sound.h"
#include "../praat.github.io/melder/melder.h"

using namespace Rcpp;

// Forward declarations
extern void NUMmachar();

// ============================================================================
// Free Functions for KlattGrid Creation
// ============================================================================

// Create empty KlattGrid
XPtr<structKlattGrid> klattgrid_create(
    double tmin,
    double tmax,
    int numberOfFormants,
    int numberOfNasalFormants,
    int numberOfNasalAntiFormants,
    int numberOfTrachealFormants,
    int numberOfTrachealAntiFormants,
    int numberOfFricationFormants,
    int numberOfDeltaFormants
) {
    NUMmachar();
    
    try {
        autoKlattGrid kg = KlattGrid_create(
            tmin, tmax,
            numberOfFormants,
            numberOfNasalFormants,
            numberOfNasalAntiFormants,
            numberOfTrachealFormants,
            numberOfTrachealAntiFormants,
            numberOfFricationFormants,
            numberOfDeltaFormants
        );
        
        if (!kg) {
            Rcpp::stop("KlattGrid_create returned null");
        }
        
        return create_xptr_from_auto<structKlattGrid>(kg);
    } catch (MelderError) {
        conststring32 err = Melder_getError();
        std::string errmsg = err ? Melder_peek32to8(err) : "Unknown error";
        Melder_clearError();
        Rcpp::stop("KlattGrid creation failed: " + errmsg);
    }
}

// Create KlattGrid from vowel parameters
XPtr<structKlattGrid> klattgrid_create_from_vowel(
    double duration,
    double f0start,
    double f1, double b1,
    double f2, double b2,
    double f3, double b3,
    double f4,
    double bandWidthFraction,
    double formantFrequencyInterval
) {
    NUMmachar();
    
    try {
        autoKlattGrid kg = KlattGrid_createFromVowel(
            duration, f0start,
            f1, b1, f2, b2, f3, b3, f4,
            bandWidthFraction,
            formantFrequencyInterval
        );
        
        if (!kg) {
            Rcpp::stop("KlattGrid_createFromVowel returned null");
        }
        
        return create_xptr_from_auto<structKlattGrid>(kg);
    } catch (MelderError) {
        conststring32 err = Melder_getError();
        std::string errmsg = err ? Melder_peek32to8(err) : "Unknown error";
        Melder_clearError();
        Rcpp::stop("KlattGrid_createFromVowel failed: " + errmsg);
    }
}

// Create example KlattGrid
XPtr<structKlattGrid> klattgrid_create_example() {
    NUMmachar();
    
    try {
        autoKlattGrid kg = KlattGrid_createExample();
        
        if (!kg) {
            Rcpp::stop("KlattGrid_createExample returned null");
        }
        
        return create_xptr_from_auto<structKlattGrid>(kg);
    } catch (MelderError) {
        conststring32 err = Melder_getError();
        std::string errmsg = err ? Melder_peek32to8(err) : "Unknown error";
        Melder_clearError();
        Rcpp::stop("KlattGrid_createExample failed: " + errmsg);
    }
}

// ============================================================================
// RKlattGrid Module Class
// ============================================================================

class RKlattGrid {
public:
    Rcpp::XPtr<structKlattGrid> ptr;
    
    // Constructors
    RKlattGrid() : ptr(R_NilValue) {}
    RKlattGrid(Rcpp::XPtr<structKlattGrid> p) : ptr(p) {}
    
    // Validation
    bool is_valid() const {
        return ptr.get() != nullptr;
    }
    
    // Time domain properties
    double get_xmin() const {
        if (!is_valid()) Rcpp::stop("Invalid KlattGrid");
        return ptr->xmin;
    }
    
    double get_xmax() const {
        if (!is_valid()) Rcpp::stop("Invalid KlattGrid");
        return ptr->xmax;
    }
    
    double get_duration() const {
        return get_xmax() - get_xmin();
    }
    
    // Synthesis to Sound
    Rcpp::XPtr<structSound> to_sound() {
        if (!is_valid()) Rcpp::stop("Invalid KlattGrid");
        
        try {
            autoSound sound = KlattGrid_to_Sound(ptr.get());
            
            if (!sound) {
                Rcpp::stop("KlattGrid_to_Sound returned null");
            }
            
            return create_xptr_from_auto<structSound>(sound);
        } catch (MelderError) {
            conststring32 err = Melder_getError();
            std::string errmsg = err ? Melder_peek32to8(err) : "Unknown error";
            Melder_clearError();
            Rcpp::stop("to_sound failed: " + errmsg);
        }
    }
    
    // Synthesis - phonation only
    Rcpp::XPtr<structSound> to_sound_phonation() {
        if (!is_valid()) Rcpp::stop("Invalid KlattGrid");
        
        try {
            autoSound sound = KlattGrid_to_Sound_phonation(ptr.get());
            
            if (!sound) {
                Rcpp::stop("KlattGrid_to_Sound_phonation returned null");
            }
            
            return create_xptr_from_auto<structSound>(sound);
        } catch (MelderError) {
            conststring32 err = Melder_getError();
            std::string errmsg = err ? Melder_peek32to8(err) : "Unknown error";
            Melder_clearError();
            Rcpp::stop("to_sound_phonation failed: " + errmsg);
        }
    }
    
    // Pitch manipulation
    double get_pitch_at_time(double t) {
        if (!is_valid()) Rcpp::stop("Invalid KlattGrid");
        return KlattGrid_getPitchAtTime(ptr.get(), t);
    }
    
    void add_pitch_point(double t, double value) {
        if (!is_valid()) Rcpp::stop("Invalid KlattGrid");
        
        try {
            KlattGrid_addPitchPoint(ptr.get(), t, value);
        } catch (MelderError) {
            Melder_clearError();
            Rcpp::stop("add_pitch_point failed");
        }
    }
    
    void remove_pitch_points(double t1, double t2) {
        if (!is_valid()) Rcpp::stop("Invalid KlattGrid");
        
        try {
            KlattGrid_removePitchPoints(ptr.get(), t1, t2);
        } catch (MelderError) {
            Melder_clearError();
            Rcpp::stop("remove_pitch_points failed");
        }
    }
    
    // Voicing amplitude
    double get_voicing_amplitude_at_time(double t) {
        if (!is_valid()) Rcpp::stop("Invalid KlattGrid");
        return KlattGrid_getVoicingAmplitudeAtTime(ptr.get(), t);
    }
    
    void add_voicing_amplitude_point(double t, double value) {
        if (!is_valid()) Rcpp::stop("Invalid KlattGrid");
        
        try {
            KlattGrid_addVoicingAmplitudePoint(ptr.get(), t, value);
        } catch (MelderError) {
            Melder_clearError();
            Rcpp::stop("add_voicing_amplitude_point failed");
        }
    }
    
    // Formant frequency manipulation (formantType: 0=oral, 1=nasal, etc.)
    double get_formant_at_time(int formantType, int iformant, double t) {
        if (!is_valid()) Rcpp::stop("Invalid KlattGrid");
        return KlattGrid_getFormantAtTime(
            ptr.get(),
            static_cast<kKlattGridFormantType>(formantType),
            iformant,
            t
        );
    }
    
    void add_formant_point(int formantType, int iformant, double t, double value) {
        if (!is_valid()) Rcpp::stop("Invalid KlattGrid");
        
        try {
            KlattGrid_addFormantPoint(
                ptr.get(),
                static_cast<kKlattGridFormantType>(formantType),
                iformant,
                t,
                value
            );
        } catch (MelderError) {
            Melder_clearError();
            Rcpp::stop("add_formant_point failed");
        }
    }
    
    void remove_formant_points(int formantType, int iformant, double t1, double t2) {
        if (!is_valid()) Rcpp::stop("Invalid KlattGrid");
        
        try {
            KlattGrid_removeFormantPoints(
                ptr.get(),
                static_cast<kKlattGridFormantType>(formantType),
                iformant,
                t1,
                t2
            );
        } catch (MelderError) {
            Melder_clearError();
            Rcpp::stop("remove_formant_points failed");
        }
    }
    
    // Bandwidth manipulation
    double get_bandwidth_at_time(int formantType, int iformant, double t) {
        if (!is_valid()) Rcpp::stop("Invalid KlattGrid");
        return KlattGrid_getBandwidthAtTime(
            ptr.get(),
            static_cast<kKlattGridFormantType>(formantType),
            iformant,
            t
        );
    }
    
    void add_bandwidth_point(int formantType, int iformant, double t, double value) {
        if (!is_valid()) Rcpp::stop("Invalid KlattGrid");
        
        try {
            KlattGrid_addBandwidthPoint(
                ptr.get(),
                static_cast<kKlattGridFormantType>(formantType),
                iformant,
                t,
                value
            );
        } catch (MelderError) {
            Melder_clearError();
            Rcpp::stop("add_bandwidth_point failed");
        }
    }
    
    // File I/O
    void save(std::string path) {
        if (!is_valid()) Rcpp::stop("Invalid KlattGrid");
        
        try {
            autoMelderString fullPath;
            MelderString_copy(&fullPath, Melder_peek8to32(path.c_str()));
            structMelderFile file {};
            Melder_pathToFile(fullPath.string, &file);
            Data_writeToTextFile(ptr.get(), &file);
        } catch (MelderError) {
            Melder_clearError();
            Rcpp::stop("save failed");
        }
    }
};

// ============================================================================
// Module Registration
// ============================================================================

RCPP_MODULE(klattgrid_module) {
    class_<RKlattGrid>("RKlattGrid")
        // Constructors
        .constructor()
        .constructor<XPtr<structKlattGrid>>()
        
        // Validation
        .method("is_valid", &RKlattGrid::is_valid)
        
        // Time domain
        .method("get_xmin", &RKlattGrid::get_xmin)
        .method("get_xmax", &RKlattGrid::get_xmax)
        .method("get_duration", &RKlattGrid::get_duration)
        
        // Synthesis
        .method("to_sound", &RKlattGrid::to_sound)
        .method("to_sound_phonation", &RKlattGrid::to_sound_phonation)
        
        // Pitch
        .method("get_pitch_at_time", &RKlattGrid::get_pitch_at_time)
        .method("add_pitch_point", &RKlattGrid::add_pitch_point)
        .method("remove_pitch_points", &RKlattGrid::remove_pitch_points)
        
        // Voicing
        .method("get_voicing_amplitude_at_time", &RKlattGrid::get_voicing_amplitude_at_time)
        .method("add_voicing_amplitude_point", &RKlattGrid::add_voicing_amplitude_point)
        
        // Formants
        .method("get_formant_at_time", &RKlattGrid::get_formant_at_time)
        .method("add_formant_point", &RKlattGrid::add_formant_point)
        .method("remove_formant_points", &RKlattGrid::remove_formant_points)
        
        // Bandwidths
        .method("get_bandwidth_at_time", &RKlattGrid::get_bandwidth_at_time)
        .method("add_bandwidth_point", &RKlattGrid::add_bandwidth_point)
        
        // File I/O
        .method("save", &RKlattGrid::save)
    ;
    
    // Factory functions
    function("klattgrid_create", &klattgrid_create);
    function("klattgrid_create_from_vowel", &klattgrid_create_from_vowel);
    function("klattgrid_create_example", &klattgrid_create_example);
}

File: R/klattgrid-module.R
# klattgrid-module.R
# R wrapper for KlattGrid Rcpp module
# Phase 2.3 - Klatt formant synthesizer for articulatory speech synthesis

#' Create a KlattGrid object
#'
#' A KlattGrid is a speech synthesizer based on the Klatt formant synthesizer.
#' It allows detailed control over phonation, vocal tract resonances, frication,
#' and other articulatory parameters.
#'
#' @param tmin Start time in seconds
#' @param tmax End time in seconds
#' @param numberOfFormants Number of oral formants (typically 6)
#' @param numberOfNasalFormants Number of nasal formants (typically 1)
#' @param numberOfNasalAntiFormants Number of nasal antiformants (typically 1)
#' @param numberOfTrachealFormants Number of tracheal formants (typically 1)
#' @param numberOfTrachealAntiFormants Number of tracheal antiformants
#  (typically 1)
#' @param numberOfFricationFormants Number of frication formants (typically 6)
#' @param numberOfDeltaFormants Number of delta formants (typically 1)
#'
#' @return KlattGrid object with S3 class
#' @export
#'
#' @examples
#' \donttest{
#' # Create empty KlattGrid
#' kg <- KlattGrid(0, 1, numberOfFormants = 6)
#'
#' # Set pitch contour and voicing amplitude (both required for synthesis)
#' kg$add_pitch_point(0.5, 100)      # 100 Hz at 0.5s
#' kg$add_voicing_amplitude_point(0.5, 90)
#'
#' # Set oral formant frequencies (formantType 1 = ORAL)
#' kg$add_formant_point(1, 1, 0.5, 500)  # F1 = 500 Hz
#' kg$add_formant_point(1, 2, 0.5, 1500) # F2 = 1500 Hz
#'
#' # Synthesize
#' sound <- kg$to_sound()
#' }
#'
#' @name KlattGrid
NULL

# ============================================================================
# Shared Method Dispatch Table
# ============================================================================

.klattgrid_methods <- new.env(hash = TRUE, parent = emptyenv())

# Validation
.klattgrid_methods$is_valid <- function(.self) .self$.cpp$is_valid()

# Time domain
.klattgrid_methods$get_xmin <- function(.self) .self$.cpp$get_xmin()
.klattgrid_methods$get_xmax <- function(.self) .self$.cpp$get_xmax()
.klattgrid_methods$get_duration <- function(.self) .self$.cpp$get_duration()

# Synthesis
.klattgrid_methods$to_sound <- function(.self) {
  sound_xptr <- .self$.cpp$to_sound()
  Sound(.xptr = sound_xptr)
}

.klattgrid_methods$to_sound_phonation <- function(.self) {
  sound_xptr <- .self$.cpp$to_sound_phonation()
  Sound(.xptr = sound_xptr)
}

# Pitch manipulation
.klattgrid_methods$get_pitch_at_time <- function(.self,
  t) .self$.cpp$get_pitch_at_time(t)

.klattgrid_methods$add_pitch_point <- function(.self, t, value) {
  .self$.cpp$add_pitch_point(t, value)
  invisible(.self)
}

.klattgrid_methods$remove_pitch_points <- function(.self, t1, t2) {
  .self$.cpp$remove_pitch_points(t1, t2)
  invisible(.self)
}

# Voicing amplitude
.klattgrid_methods$get_voicing_amplitude_at_time <- function(.self, t) {
  .self$.cpp$get_voicing_amplitude_at_time(t)
}

.klattgrid_methods$add_voicing_amplitude_point <- function(.self, t, value) {
  .self$.cpp$add_voicing_amplitude_point(t, value)
  invisible(.self)
}

# Formant manipulation
# formantType: 1=oral, 2=nasal, 3=frication, 4=tracheal, 5=nasal_anti,
#  6=tracheal_anti, 7=delta
.klattgrid_methods$get_formant_at_time <- function(.self, formantType,
  iformant, t) {
  .self$.cpp$get_formant_at_time(as.integer(formantType),
    as.integer(iformant), t)
}

.klattgrid_methods$add_formant_point <- function(.self, formantType, iformant,
  t, value) {
  .self$.cpp$add_formant_point(as.integer(formantType), as.integer(iformant),
    t, value)
  invisible(.self)
}

.klattgrid_methods$remove_formant_points <- function(.self, formantType,
  iformant, t1, t2) {
  .self$.cpp$remove_formant_points(as.integer(formantType),
    as.integer(iformant), t1, t2)
  invisible(.self)
}

# Bandwidth manipulation
.klattgrid_methods$get_bandwidth_at_time <- function(.self, formantType,
  iformant, t) {
  .self$.cpp$get_bandwidth_at_time(as.integer(formantType),
    as.integer(iformant), t)
}

.klattgrid_methods$add_bandwidth_point <- function(.self, formantType,
  iformant, t, value) {
  .self$.cpp$add_bandwidth_point(as.integer(formantType),
    as.integer(iformant), t, value)
  invisible(.self)
}

# File I/O
.klattgrid_methods$save <- function(.self, path) {
  .self$.cpp$save(path)
  invisible(.self)
}

# Display
.klattgrid_methods$print <- function(.self) {
  cat("KlattGrid object\n")
  cat("  Duration:", .self$.cpp$get_duration(), "s\n")
  cat("  Time range: [", .self$.cpp$get_xmin(), ",", .self$.cpp$get_xmax(),
    "]\n")
  invisible(.self)
}

lockEnvironment(.klattgrid_methods, bindings = TRUE)

# ============================================================================
# S3 Dispatch
# ============================================================================

#' @method $ KlattGrid
#' @export
`$.KlattGrid` <- function(x, name) {
  val <- .subset2(x, name)
  if (!is.null(val)) return(val)
  method <- .klattgrid_methods[[name]]
  if (is.null(method)) return(NULL)
  function(...) method(x, ...)
}

# ============================================================================
# Constructors
# ============================================================================

#' @export
KlattGrid <- function(tmin = 0.0,
                      tmax = 1.0,
                      numberOfFormants = 6L,
                      numberOfNasalFormants = 1L,
                      numberOfNasalAntiFormants = 1L,
                      numberOfTrachealFormants = 1L,
                      numberOfTrachealAntiFormants = 1L,
                      numberOfFricationFormants = 6L,
                      numberOfDeltaFormants = 1L) {
  mod <- get_module("klattgrid_module")
  xptr <- mod$klattgrid_create(
    tmin, tmax,
    as.integer(numberOfFormants),
    as.integer(numberOfNasalFormants),
    as.integer(numberOfNasalAntiFormants),
    as.integer(numberOfTrachealFormants),
    as.integer(numberOfTrachealAntiFormants),
    as.integer(numberOfFricationFormants),
    as.integer(numberOfDeltaFormants)
  )
  cpp_obj <- mod$RKlattGrid$new(xptr)

  structure(list(
    .cpp = cpp_obj
  ), class = c("KlattGrid", "PraatObject"))
}

#' Create KlattGrid from vowel parameters
#'
#' Creates a KlattGrid pre-configured for synthesizing a vowel sound
#' with specified formant frequencies and bandwidths.
#'
#' @param duration Duration in seconds
#' @param f0start Starting F0 in Hz
#' @param f1 First formant frequency in Hz
#' @param b1 First formant bandwidth in Hz
#' @param f2 Second formant frequency in Hz
#' @param b2 Second formant bandwidth in Hz
#' @param f3 Third formant frequency in Hz
#' @param b3 Third formant bandwidth in Hz
#' @param f4 Fourth formant frequency in Hz (optional)
#' @param bandWidthFraction Bandwidth as fraction of frequency (default 0.05)
#' @param formantFrequencyInterval Formant spacing interval in Hz (default 1000)
#'
#' @return KlattGrid object configured for vowel
#' @examples
#' kg <- klattgrid_create_from_vowel(duration = 0.3, f0start = 120)
#' sound <- kg$to_sound()
#' @export
klattgrid_create_from_vowel <- function(duration = 0.5,
                                      f0start = 100.0,
                                      f1 = 500.0, b1 = 50.0,
                                      f2 = 1500.0, b2 = 100.0,
                                      f3 = 2500.0, b3 = 150.0,
                                      f4 = 3500.0,
                                      bandWidthFraction = 0.05,
                                      formantFrequencyInterval = 1000.0) {
  mod <- get_module("klattgrid_module")
  xptr <- mod$klattgrid_create_from_vowel(
    duration, f0start,
    f1, b1, f2, b2, f3, b3, f4,
    bandWidthFraction,
    formantFrequencyInterval
  )
  cpp_obj <- mod$RKlattGrid$new(xptr)

  structure(list(
    .cpp = cpp_obj
  ), class = c("KlattGrid", "PraatObject"))
}

#' Create example KlattGrid
#'
#' Creates a demonstration KlattGrid with pre-configured parameters
#' for testing the synthesizer.
#'
#' @return KlattGrid example object
#' @examples
#' kg <- klattgrid_create_example()
#' sound <- kg$to_sound()
#' @export
klattgrid_create_example <- function() {
  mod <- get_module("klattgrid_module")
  xptr <- mod$klattgrid_create_example()
  cpp_obj <- mod$RKlattGrid$new(xptr)

  structure(list(
    .cpp = cpp_obj
  ), class = c("KlattGrid", "PraatObject"))
}

# ============================================================================
# Deprecated aliases — forward to new names with a deprecation warning
# ============================================================================

#' @rdname klattgrid_create_from_vowel
#' @usage # Deprecated: use klattgrid_create_from_vowel() instead
#' @export
KlattGrid_createFromVowel <- function(duration = 0.5,
                                      f0start = 100.0,
                                      f1 = 500.0, b1 = 50.0,
                                      f2 = 1500.0, b2 = 100.0,
                                      f3 = 2500.0, b3 = 150.0,
                                      f4 = 3500.0,
                                      bandWidthFraction = 0.05,
                                      formantFrequencyInterval = 1000.0) {
  .Deprecated("klattgrid_create_from_vowel")
  klattgrid_create_from_vowel(duration, f0start, f1, b1, f2, b2, f3, b3, f4,
                               bandWidthFraction, formantFrequencyInterval)
}

#' @rdname klattgrid_create_example
#' @usage # Deprecated: use klattgrid_create_example() instead
#' @export
KlattGrid_createExample <- function() {
  .Deprecated("klattgrid_create_example")
  klattgrid_create_example()
}

# ============================================================================
# S3 Methods
# ============================================================================
